src/vm/analysis/cost_counter/natives.rs

```diff
@@ -0,0 +1,589 @@
+                runtime: Linear(constants::DB_WRITE_RUNTIME_COST_A, constants::DB_WRITE_RUNTIME_COST_B),
+            }.compute_cost(value_size)?;
+
+            hash_cost.add(&write_cost)?;
+
+            // the cost of computing the value arg
+            hash_cost.add(&inst.handle_expression(value_arg)?)?;
+
+            Ok(hash_cost)
+        },
+        SetEntry | InsertEntry => {
+            assert!(args.len() == 3);
+            let map_name = args[0].match_atom().expect("Argument should be atomic name.");
+            let key_tuple = &args[1];
+            let value_tuple = &args[2];
+
+            let value_size =  u64::from(inst.type_map.get_type(value_tuple)
+                                        .expect(TYPE_ANNOTATED_FAIL)
+                                        .size());
+
+            let key_size =  u64::from(inst.type_map.get_type(key_tuple)
+                                      .expect(TYPE_ANNOTATED_FAIL)
+                                      .size());
+            // the cost of the hash lookup...
+            let mut hash_cost = get_hash_cost(map_name.len() as u64, key_size)?;
+
+            // the cost of the database op...
+            let write_cost = SimpleCostSpecification {
+                read_count: Constant(0), read_length: Constant(0),
+                write_length: Constant(value_size), write_count: Constant(1),
+                runtime: Linear(constants::DB_WRITE_RUNTIME_COST_A, constants::DB_WRITE_RUNTIME_COST_B),
+            }.compute_cost(value_size)?;
+
+            hash_cost.add(&write_cost)?;
+
+            // the cost of computing the key, value args
+            hash_cost.add(&handle_tuple_argument(inst, key_tuple)?)?;
+            hash_cost.add(&handle_tuple_argument(inst, value_tuple)?)?;
+
+            Ok(hash_cost)
+        },
+        DeleteEntry => {
+            assert!(args.len() == 2);
+            let map_name = args[0].match_atom().expect("Argument should be atomic name.");
+            let key_tuple = &args[1];
+
+            let key_size =  u64::from(inst.type_map.get_type(key_tuple)
+                                      .expect(TYPE_ANNOTATED_FAIL)
+                                      .size());
+            // the cost of the hash lookup...
+            let mut hash_cost = get_hash_cost(map_name.len() as u64, key_size)?;
+
+            // the cost of the database op...
+            let write_cost = SimpleCostSpecification {
+                read_count: Constant(1), read_length: Constant(0),
+                write_length: Constant(1), write_count: Constant(1),
+                runtime: Linear(constants::DB_WRITE_RUNTIME_COST_A, constants::DB_WRITE_RUNTIME_COST_B),
+            }.compute_cost(1)?;
+
+            hash_cost.add(&write_cost)?;
+
+            // the cost of computing the key, value args
+            hash_cost.add(&handle_tuple_argument(inst, key_tuple)?)?;
+
+            Ok(hash_cost)
+        },
+        MintAsset => {
```

Comment:
When we're doing a lookup of an asset name, we are building a key that includes the contract-identifier (principal+contract-name). Should that be included in the `get_hash_cost` for the functions `MintAsset`, `TransferAsset`, `GetAssetOwner` ?

Reply:
Yes, it needs to be. But we can set that to a constant, since the maximum contract name length is 128 characters.